ENH: accept dict of column:dtype as dtype argument in DataFrame.astype

StephenKappel opened the pull request:
Continued in https://github.com/pydata/pandas/pull/13375

---

closes #7271

By passing a dict of {column name/column index: dtype}, multiple columns can be cast to different data types in a single command. Now users can do:

`df = df.astype({'my_bool', 'bool', 'my_int': 'int'})`

or:

`df = df.astype({0, 'bool', 1: 'int'})`

instead of:

```
df['my_bool'] = df.my_bool.astype('bool')
df['my_int'] = df.my_int.astype('int')
```

wesm commented on pandas/core/frame.py:
```diff
@@ -2305,6 +2305,35 @@ def is_dtype_instance_mapper(column, dtype):
+        -------
+        casted : type of caller
+        """
+        if type(dtype) == dict:
```
use `isinstance` here to get subclasses, too

MaximilianR replied:
`isinstance(dtype, collections.Mapping)` if you want to be really precise

jreback commented on pandas/core/frame.py:
```diff
@@ -2305,6 +2305,35 @@ def is_dtype_instance_mapper(column, dtype):
+        if isinstance(dtype, collections.Mapping):
+            df = DataFrame(data=self, copy=copy, **kwargs)
+            for col_key, col_dtype in dtype.items():
+                df.ix[:, col_key] = df.ix[:, col_key].astype(col_dtype)
```
this iterative updating will be quite inefficient as it will potentially copy the entire new dtype each time as it grows; this should change with the internals refactor #11970 , but for now do something like:

```
astyped_cols = pd.concat([ df[col_key].astype(col_dtype) for col_key, col_dtype in dtype.items() ], axis=1)
return pd.concat([astyped_cols, df[[df.columns.difference(dtype.keys())]], axis=1).reindex(columns=df.columns)
```

which should minimize the copying.

jreback commented:
astype currently has the ``copy`` kw. This is a big bogus as this almost always actually does copy if its changing to a new dtype (the exception is when its going to the same dtype).

related to #5769. It think we should change this to ``inplace`` (and deprecate ``copy``) as its a bit more clear whats actually going on.

This could be done in another PR (before this one), or in this one is ok too.

jreback commented:
further I wouldn't use ``.ix`` here as that would interpret numeric values as positional indexers for columns as this is a potentially confusing situation (and we don't do this anywhere else). If a non-matching column is passed then it will simply raise a ``KeyError``. Instead use ``[]``.

So pls add some more tests, including some invalid columns.

jreback commented on pandas/core/frame.py:
```diff
@@ -2305,6 +2305,35 @@ def is_dtype_instance_mapper(column, dtype):
+                df.ix[:, col_key] = df.ix[:, col_key].astype(col_dtype)
+            return df
+        else:
+            return super(DataFrame, self).astype(dtype=dtype, copy=copy,
```
you can simply return instead of using ``else`` here.

jreback commented on pandas/core/frame.py:
```diff
@@ -2305,6 +2305,35 @@ def is_dtype_instance_mapper(column, dtype):
+        """
+        if isinstance(dtype, collections.Mapping):
+            df = DataFrame(data=self, copy=copy, **kwargs)
+            for col_key, col_dtype in dtype.items():
```
you need to pass ``copy`` and ``raise_no_error`` to the ``.astype`` as well. (but see my note about ``copy`` below)

jreback commented:
@StephenKappel can you rebase/update

sinhrks committed:
BUG: Series ops with object dtype may incorrectly fail

closes #13043
closes #13072

jreback commented:
@StephenKappel can you rebase / update

StephenKappel commented:
Sorry for neglecting this. The semester is finally over, so I'll get this cleaned up today :-)

StephenKappel committed:
ENH: inplace dtype changes, df per-column dtype changes; GH7271

StephenKappel committed:
ENH: inplace dtype changes, df per-column dtype changes; GH7271

StephenKappel commented:
@jreback - Rebased and updated based on your earlier feedback. Please let me know if any further updates are needed.

jreback commented on pandas/core/frame.py:
```diff
@@ -3772,6 +3772,47 @@ def update(self, other, join='left', overwrite=True, filter_func=None,
+        Parameters
+        ----------
+        dtype : numpy.dtype or Python type (to cast entire DataFrame to the
+            same type). Alternatively, {col: dtype, ...}, where col is a column
```
why are you adding a method her? this should all be done in ``generic.py``

jreback commented on pandas/core/generic.py:
```diff
@@ -2930,22 +2930,30 @@ def blocks(self):
+            new_data = self._data.astype(dtype=dtype, copy=False,
+                                         raise_on_error=raise_on_error,
+                                         **kwargs)
+            self._update_inplace(new_data)
```
verbose, you can do this like:

```
new_data = ....... copy=not inplace,...

if inplace:
    self._update_inplace(new_data)
else:
    return self._constructor(....)
```

jreback commented on pandas/tests/frame/test_dtypes.py:
```diff
@@ -372,6 +372,76 @@ def test_astype_str(self):
             expected = DataFrame(['1.12345678901'])
             assert_frame_equal(result, expected)
 
+    def test_astype_dict(self):
```
you can test here, but test for ``Series`` as well (e.g. it would have to be a 1-element dictionary)

StephenKappel replied:
Is there a more elegant way to handle the `Series` case than just adding a new `if` statement to handle the `Series` explicitly? The above implementation doesn't work for `Series` because the `df[label]` accessor would try to access rows in the `Series`, and I think we want to map the keys in the dict to the `Series.name` property.

jreback replied:
you can just test its ok, e.g. if ``self.ndim == 1``

jreback commented on pandas/tests/frame/test_dtypes.py:
```diff
@@ -372,6 +372,76 @@ def test_astype_str(self):
+            'b': Series(['0', '1', '2', '3', '4']),
+            'c': c,
+            'd': Series([1.0, 2.0, 3.14, 4.0, 5.4], dtype='float32')})
+        astyped = df.astype({'b': 'str', 'd': 'float32'})
```
```
result = ...
expected = DataFrame(.....)
assert_frame_equal(result, expected)
```

jreback commented on pandas/tests/frame/test_dtypes.py:
```diff
@@ -372,6 +372,76 @@ def test_astype_str(self):
+        assert_frame_equal(astyped, expected)
+        assert_frame_equal(df, original)
+        self.assertEqual(astyped.b.dtype, 'object')
+        self.assertEqual(astyped.d.dtype, 'float32')
```
don't use assertEqual like this, simply use ``assert_frame_equal``

jreback commented on pandas/tests/frame/test_dtypes.py:
```diff
@@ -372,6 +372,76 @@ def test_astype_str(self):
+        assert_frame_equal(df, expected)
+        df.astype({'b': np.float32, 'c': 'float32', 'd': np.float32},
+                  inplace=True)
+        self.assertEqual(df.a.dtype, original.a.dtype)
```
again construct an expected

jreback commented on pandas/tests/series/test_dtypes.py:
```diff
@@ -133,6 +133,19 @@ def test_astype_unicode(self):
+    def test_astype_inplace(self):
+        s = Series(np.random.randn(5), name='foo')
+
+        for dtype in ['float32', 'float64', 'int64', 'int32']:
```
``{s.name : dtype}`` would be valid as well

StephenKappel committed:
ENH: NDFrame astype() now accepts inplace arg and dtype arg can be a mapping of col to type; GH7271

StephenKappel committed:
ENH: NDFrame astype() now accepts inplace arg and dtype arg can be a mapping of col to type; GH7271

jreback commented on pandas/core/generic.py:
```diff
@@ -2930,25 +2931,61 @@ def blocks(self):
-        dtype : numpy.dtype or Python type
+        dtype : numpy.dtype or Python type (to cast entire DataFrame to the
+            same type). Alternatively, {col: dtype, ...}, where col is a column
+            label and dtype is a numpy.dtype or Python type (to cast one or
```
use PandasObject rather than DataFrame.

add versionadded 0.18.2 tag

jorisvandenbossche replied:
@jreback `PandasObject` is not really meaningful towards users. But a more general "pandas object" (no code quoting) is maybe a possibility?

Further, @StephenKappel, can you format this according to the numpydoc standard? (https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections):

```
param : type
    Description of parameter.
```

jreback commented on pandas/core/generic.py:
```diff
@@ -2930,25 +2931,61 @@ def blocks(self):
+            more of the DataFrame's columns to column-specific types).
+        copy : deprecated; use inplace instead
+        inplace : boolean, default False
+            Modify the NDFrame in place (do not create a new object)
```
version added tag

jreback replied:
actually I don't think we can have an ``inplace`` flag here. you can't really do this except in some extreme circumstances (e.g. you *can* inplace ints to floats), but its not worth supporting at all and is quite convoluted.

The copy flag was just to prevent copies if we are already the same dtype. SO let's leave all of that alone.

jreback commented on pandas/core/generic.py:
```diff
@@ -2930,25 +2931,61 @@ def blocks(self):
+            Modify the NDFrame in place (do not create a new object)
         raise_on_error : raise on invalid input
-        kwargs : keyword arguments to pass on to the constructor
+        kwargs : keyword arguments to pass on to the constructor if
```
why would this only work if inplace=False?

jreback commented on pandas/core/generic.py:
```diff
@@ -2930,25 +2931,61 @@ def blocks(self):
+        if isinstance(dtype, collections.Mapping):
+            if self.ndim == 1:  # i.e. Series
+                if len(dtype) > 1 or list(dtype.keys())[0] != self.name:
+                    if raise_on_error:
```
you will always raise here its a specification issue

jreback commented on pandas/core/generic.py:
```diff
@@ -2930,25 +2931,61 @@ def blocks(self):
+                        raise KeyError('Only the Series name can be used for '
+                                       'the key in Series dtype mappings.')
+                    return
+                for key, value in dtype.items():
```
``dtype = dtype.values()[0]``

jreback commented on pandas/core/generic.py:
```diff
@@ -2930,25 +2931,61 @@ def blocks(self):
+            from pandas.tools.merge import concat
+            casted_cols = [self[col].astype(typ, copy=copy)
+                           for col, typ in dtype.items()]
+            other_col_labels = self.columns.difference(dtype.keys())
```
do the astype on each column, NOT inplace. Then assemble them. Then if its inplace update, else return it.

jreback committed:
DOC: xref #13112, add back lexsorting example

tacaswell committed:
BUG: mpl fix to AutoDatFromatter to fix second/us-second formatters

revert part of #11770
xref: https://github.com/matplotlib/matplotlib/issues/6365

closes #13131

The mistake in #11770 was missing that pandas had a 1/us not 1/s
scaled bucket.

jreback committed:
COMPAT: boto import issues

Author: Jeff Reback <jeff@reback.net>

Closes #13136 from jreback/boto and squashes the following commits:

dcb02d2 [Jeff Reback] COMPAT: boto import issues

yaduart committed:
BUG: Added checks for NaN in __call__ of EngFormatter

closes #11981

Author: Yadunandan <yadunandan2001@gmail.com>

Closes #13124 from yaduart/bugfix-11981 and squashes the following commits:

8de1f64 [Yadunandan] BUG: Added checks for Nan in __call__ of EngFormatter

sinhrks committed:
TST: fix assert_categorical_equal message

stage 1 of #13076

Author: sinhrks <sinhrks@gmail.com>

Closes #13080 from sinhrks/test_categorical_message and squashes the following commits:

81172ce [sinhrks] TST: fix assert_categorical_equal message

jreback committed:
PERF: quantile now operates per block boosting perf / fix quantile with nan

closes #11623
closes #13098

Author: Jeff Reback <jeff@reback.net>

Closes #13122 from jreback/quantile and squashes the following commits:

aad72cb [Jeff Reback] PERF: quantile now operates per block boosting perf REGR: series quantile with nan

dsm054 committed:
DOC: Fix delim_whitespace regex typo.

Minor typo in the explanation of delim_whitespace which tripped up a
user on SO (although the user should probably have been using
`delim_whitespace=True` directly anyhow.)

Author: dsm054 <dsm054@gmail.com>

Closes #13165 from dsm054/fix-delim_whitespace-regex and squashes the following commits:

c8f13d2 [dsm054] DOC: Fix delim_whitespace regex typo.

gliptak committed:
BUG: Correct KeyError from matplotlib when processing Series yerr

closes #11858

Author: Gábor Lipták <gliptak@gmail.com>

Closes #13114 from gliptak/yerr1 and squashes the following commits:

926329a [Gábor Lipták] Correct KeyError from matplotlib when processing Series xerr/yerr

sinhrks committed:
BUG: Misc fixes for SparseSeries indexing with MI

closes #13144

Author: sinhrks <sinhrks@gmail.com>

Closes #13163 from sinhrks/sparse_multi and squashes the following commits:

eb24102 [sinhrks] BUG: Misc fixes for SparseSeries indexing with MI

sinhrks committed:
ENH/BUG: str.extractall doesn't support index

closes #10008

Author: sinhrks <sinhrks@gmail.com>

Closes #13156 from sinhrks/str_extractall and squashes the following commits:

ed854ef [sinhrks] ENH/BUG: str.extractall doesn't support index

Xndr7 committed:
DOC: Fix additional join examples in "10 Minutes to pandas" #13029

 - [x] closes #13029

Author: Sanjiv Lobo <sldesalobo@gmail.com>

Closes #13171 from Xndr7/fix-additional-join-examples-in-"10-Minutes-to-pandas"-#13029 and squashes the following commits:

633c7ff [Sanjiv Lobo] fixed docs for issue #13029

jreback committed:
BUG: Bug in .groupby(..).resample(..) when the same object is called multiple times

closes #13174

Author: Jeff Reback <jeff@reback.net>

Closes #13175 from jreback/resample and squashes the following commits:

56b405e [Jeff Reback] BUG: Bug in .groupby(..).resample(..) when the same object is called multiple times

gfyoung committed:
DOC: Clarify Categorical Crosstab Behaviour

Follow-on to #13073 by explaining the `Categorical` behaviour in the
documentation.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13177 from gfyoung/crosstab-categorical-explain and squashes the following commits:

11ebb94 [gfyoung] DOC: Clarify Categorical Crosstab Behaviour

kawochen committed:
BUG: GH12896 where extra elements are returned in MultiIndex slicing

closes #12896

Author: Ka Wo Chen <kawoc@tepper.cmu.edu>

Closes #13117 from kawochen/BUG-FIX-12896 and squashes the following commits:

7d49346 [Ka Wo Chen] BUG: GH12896 where extra elements are returned in MultiIndex slicing

neirbowj committed:
TST: Use compatible time zones

xref #13186
closes #13190

quintusdias committed:
COMPAT: Add Pathlib, py.path support for read_hdf

Closes #11773

Author: John Evans <john.g.evans.ne@gmail.com>

Closes #12930 from quintusdias/issue11773 and squashes the following commits:

dcee282 [John Evans] COMPAT: Add Pathlib, py.path support for read_hdf, to_hdf

jreback committed:
COMPAT/TST: sparse formatting test for platform, xref #13163

max-sixty committed:
CLN: no return on init

Author: Maximilian Roos <maximilian@sixtycapital.com>

Closes #13197 from MaximilianR/init-return and squashes the following commits:

ee5072a [Maximilian Roos] formatting

starplanet committed:
BUG: fix to_records confict with unicode_literals #13172

closes #13172

Author: zhangjinjie <zhangjinjie@yimian.com.cn>

Closes #13178 from starplanet/bugfix-13172 and squashes the following commits:

62b00d3 [zhangjinjie] BUG: fix to_records confict with unicode_literals #13172

sinhrks committed:
BUG: Period and Series/Index comparison raises TypeError

Author: sinhrks <sinhrks@gmail.com>

Closes #13200 from sinhrks/period_comp and squashes the following commits:

aadf669 [sinhrks] BUG: Period and Series/Index comparison raises TypeError

jreback committed:
TST: change test comparison to work on older numpies, #13178

chris-b1 committed:
PERF: DataFrame transform

closes #12737
closes #13191

Author: Chris <cbartak@gmail.com>

Closes #13192 from chris-b1/transform-perf and squashes the following commits:

0af1e55 [Chris] revert casting logic
d61d4e0 [Chris] handle duplicate column case
9d78f65 [Chris] other categorical test name fix
045d0c7 [Chris] add back some casting
b66a1c8 [Chris] PERF: DataFrame transform

sinhrks committed:
BUG: Sparse creation with object dtype may raise TypeError

closes #11633
closes #11856

Author: sinhrks <sinhrks@gmail.com>

Closes #13201 from sinhrks/sparse_isnull and squashes the following commits:

443b47e [sinhrks] BUG: Sparse creation with object dtype may raise TypeError

Commit:
TST: Test resampling with NaT

closes #13020

Author: Alex Alekseyev <alex.alekseyev@twosigma.com>

Closes #13164 from evectant/issue-13020 and squashes the following commits:

c7e6f5f [Alex Alekseyev] TST: Test resampling with NaT

Commit:
BUG: Fix #13213 json_normalize() and non-ascii characters in keys

closes #13213

Author: Felix Marczinowski <felix.marczinowski@blue-yonder.com>

Closes #13214 from fmarczin/13213-unicode-json_normalize and squashes the following commits:

22e01b2 [Felix Marczinowski] fix linter warnings
44745ca [Felix Marczinowski] fix tests for py3
25fd0f8 [Felix Marczinowski] move test, fix py3 issue
7a38110 [Felix Marczinowski] add whatsnew note
dd7302c [Felix Marczinowski] remove encoding signature from test
4dcd2c5 [Felix Marczinowski] fix for #13213
b9751e9 [Felix Marczinowski] add test for #13213

eddiejessup committed:
BUG: Fix argument order in call to super

Author: Elliot Marsden <elliot.marsden@gmail.com>

Closes #12924 from eddiejessup/master and squashes the following commits:

b495e32 [Elliot Marsden] BUG: Fix argument order in call to super

jreback committed:
DOC: add v0.19.0 whatsnew doc

hassanshamim committed:
BUG: GH13219 Fixed. Allow unicode values in usecols

closes #13219
closes #13233

gfyoung committed:
COMPAT: Further Expand Compatibility with fromnumeric.py

Follow-on to #12810 by expanding compatibility with fromnumeric.py
in the following modules:
  1) tslib.pyx
  2) window.py
  3) groupby.py and resample.py (shared classes)

Closes #12811.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13148 from gfyoung/fromnumeric-compat-continued and squashes the following commits:

eb4762c [gfyoung] COMPAT: Expand compatibility with fromnumeric.py

jreback committed:
BUG: Bug in .to_datetime() when passing integers or floats, no unit and errors=coerce

closes #13180

Author: Jeff Reback <jeff@reback.net>

Closes #13183 from jreback/coerce and squashes the following commits:

0076151 [Jeff Reback] BUG: Bug in .to_datetime() when passing integers or floats, no unit and errors=coerce

adneu committed:
BUG: GH12824 fixed apply() returns different result depending on whet…

closes #12824

Author: adneu <aneumann31@gmail.com>

Closes #12977 from adneu/GH12824 and squashes the following commits:

9cf7e01 [adneu] BUG: GH12824 fixed apply() returns different result depending on whether first result is None

thejohnfreeman committed:
API/COMPAT: add pydatetime-style positional args to Timestamp constructor

closes #10758
closes #11630

Author: John Freeman <jfreeman08@gmail.com>

Closes #12482 from thejohnfreeman/GH-10758 and squashes the following commits:

79d63d4 [John Freeman] Format comments, test for NaT
0ac786f [John Freeman] Special object for missing first argument
a334eab [John Freeman] PEP8
9c1e2dc [John Freeman] review fixes: versionadded, issue links, repr tests, legacy
6fad30b [John Freeman] Fix docstring; add tests for kwargs
5c34c04 [John Freeman] Support positional and keyword arguments for Timestamp
0d6884b [John Freeman] API/COMPAT: add pydatetime-style positional args to Timestamp constructor

jreback committed:
TST: xref #13183, for windows compat

gfyoung committed:
Reverse numpy compat changes to tslib.pyx

Title is self-explanatory.  xref #13148.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13246 from gfyoung/tslib-compat-undo and squashes the following commits:

66160f9 [gfyoung] Reverse numpy compat changes to tslib.pyx

max-sixty committed:
BUG: Empty PeriodIndex issues

closes #13067
closes #13212

Author: Maximilian Roos <maximilian@sixtycapital.com>

Closes #13079 from MaximilianR/period_resample_0 and squashes the following commits:

8c7b9db [Maximilian Roos] empty PeriodIndex issues

Commit:
API: Use np.random's RandomState when seed is None in .sample

closes #13143

Author: Allen Riddell <abr@ariddell.org>

Closes #13161 from ariddell/feature/sample-numpy-random-seed and squashes the following commits:

595b0bc [Allen Riddell] Use np.random's RandomState when seed is None

sinhrks committed:
TST: check internal Categorical

 - [x] closes #13076   - [x] tests added / passed   - [x] passes ``git
diff upstream/master | flake8 --diff``

Author: sinhrks <sinhrks@gmail.com>

Closes #13249 from sinhrks/test_categorical and squashes the following commits:

f536644 [sinhrks] TST: check internal Categorical

sinhrks committed:
TST/ERR: Add Period ops tests / fix error message

xref to #13242
closes #13251

Author: sinhrks <sinhrks@gmail.com>

Closes #13250 from sinhrks/period_test and squashes the following commits:

33a04f3 [sinhrks] TST/ERR: Add Period ops tests / fix error message

Commit:
ENH: support decimal option in PythonParser #12933

closes #12933

Author: Camilo Cota <ccota@riplife.es>

Closes #13189 from camilocot/12933 and squashes the following commits:

465272e [Camilo Cota] Benchmark decimal option in read_csv for c engine
9f42d0c [Camilo Cota] double backticks around decimal and engine='python'
dc8ca62 [Camilo Cota] fix test_empty_decimal_marker comment
49613fe [Camilo Cota] Assert read_csv error message in test_empty_decimal_marker
d821052 [Camilo Cota] fix test_empty_decimal_marker comment
f71509d [Camilo Cota] Include descritive what's new line
803356e [Camilo Cota] set nonnum regex in init method
1472d80 [Camilo Cota] Include the issue number in what's new
b560fda [Camilo Cota] Fix what's new
dc7acd1 [Camilo Cota] ENH: support decimal option in PythonParser #12933

gliptak committed:
ERR: Correct ValueError invalid type promotion exception

closes #12599

Author: Gábor Lipták <gliptak@gmail.com>

Closes #13234 from gliptak/invalidtypepromotion and squashes the following commits:

88f144b [Gábor Lipták] Correct ValueError invalid type promotion exception

pijucha committed:
BUG: Fix #13149 and ENH: 'copy' param in Index.astype()

closes #13149

1. Float64Index.astype(int) raises ValueError if a NaN is present.
   Previously, it converted NaN's to the smallest negative integer.
2. TimedeltaIndex.astype(int) and DatetimeIndex.astype(int) return.  Int64Index, which is consistent
   with behavior of other Indexes.  Previously, they returned a numpy.array of ints.
3. Added bool parameter 'copy' to Index.astype()
4. Fixed core.common.is_timedelta64_ns_dtype().
5. Set a default NaT representation to a string type in a parameter of
   DatetimeIndex._format_native_types().  Previously, it produced a
   unicode u'NaT' in Python2.

Author: pijucha <pi.jucha@gmail.com>

Closes #13209 from pijucha/bug13149 and squashes the following commits:

8b29902 [pijucha] BUG: Fix #13149 and ENH: 'copy' param in Index.astype()

gfyoung committed:
BUG, ENH: Add support for parsing duplicate columns

Closes #7160
Closes #9424

Author: gfyoung <gfyoung17@gmail.com>

Closes #12935 from gfyoung/dupe-col-names and squashes the following commits:

ef7636f [gfyoung] BUG, ENH: Add support for parsing duplicate columns

sinhrks committed:
TST: assert_dict_equal to check input type

Author: sinhrks <sinhrks@gmail.com>

Closes #13264 from sinhrks/test_dict and squashes the following commits:

2a7b9b1 [sinhrks] TST: assert_dict_equal to check input type

sinhrks committed:
BUG: remove_unused_categories dtype coerces to int64

Author: sinhrks <sinhrks@gmail.com>

Closes #13261 from sinhrks/categorical_remove_dtype and squashes the following commits:

b5cbe2c [sinhrks] BUG: remove_unused_categories dtype coerces to int64

jreback committed:
BUG: Bug in selection from a HDFStore with a fixed format and start and/or stop will now return the selected range

closes #8287

Author: Jeff Reback <jeff@reback.net>

Closes #13267 from jreback/stop and squashes the following commits:

39faa23 [Jeff Reback] BUG: Bug in selection from a HDFStore with a fixed format and start and/or stop specified will now return the selected range

mortada committed:
DOC: fixed typos in GroupBy document

Author: Mortada Mehyar <mortada.mehyar@gmail.com>

Closes #13270 from mortada/docs and squashes the following commits:

a3dea39 [Mortada Mehyar] DOC: fixed typos in GroupBy document

roycoding committed:
BUG: Fix for resampler for grouping kwarg bug

closes #13235
closes #13241

gfyoung committed:
BUG: Properly validate and parse nrows in read_csv

1) Allows `float` values for `nrows` for the Python engine  2)
Prevents abuse of the `nrows` argument for the CParser (e.g. you can
passing `nrows=1.2`)    Closes #10476.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13275 from gfyoung/nrows-bug-validate and squashes the following commits:

d856051 [gfyoung] BUG: Properly validate and parse nrows in read_csv

jreback committed:
COMPAT: extension dtypes (DatetimeTZ, Categorical) are now Singleton cached objects

allows for proper is / == comparisons

closes #13285

jreback commented:
can you rebase / update?

gfyoung committed:
BUG, ENH: Improve infinity parsing for read_csv

1) Allow mixed-case infinity strings for the Python engine
2) Interpret `+inf` as positive infinity for both engines

Author: gfyoung <gfyoung17@gmail.com>

Closes #13274 from gfyoung/floatify-infinity-parsing and squashes the following commits:

f37b130 [gfyoung] BUG, ENH: Improve infinity parsing in read_csv

nparley committed:
TST: Remove imp and just use importlib to avoid memory error when showing versions

closes #13282

Author: Neil Parley <mail@neilparley.co.uk>

Closes #13284 from nparley/imp and squashes the following commits:

29b98f7 [Neil Parley] Move import to top
ca5fa7a [Neil Parley] Remove imp and just use importlib to avoid memory error when importing blosc.

gliptak committed:
ERR: error in datetime conversion with non-convertibles

closes #11853

Author: Gábor Lipták <gliptak@gmail.com>

Closes #13176 from gliptak/dtbool1 and squashes the following commits:

5179d1d [Gábor Lipták] Bug in pd.to_datetime when passing bools; will now respect the errors value

edublancas committed:
DOC: Improved documentation for DataFrame.join

closes #12188

Author: Eduardo Blancas Reyes <edu.blancas@gmail.com>

Closes #12193 from edublancas/master and squashes the following commits:

a66f2ea [Eduardo Blancas Reyes] DOC: improves DataFrame.join documentation
8266cdc [Eduardo Blancas Reyes] DOC: improves DataFrame.join documentation

sinhrks committed:
TST/CLN: remove np.assert_equal

Author: sinhrks <sinhrks@gmail.com>

Closes #13263 from sinhrks/test_assert_equal and squashes the following commits:

a5b2a67 [sinhrks] TST/CLN: remove np.assert_equal

pfrcks committed:
DOC: Added an example of pitfalls when using astype

 - [x] closes #13260    - [ ] tests added / passed   - [ ] passes
``git diff upstream/master | flake8 --diff``   - [ ] whatsnew entry

Author: Amol <pfrcks@gmail.com>

Closes #13278 from pfrcks/bug13260 and squashes the following commits:

035a177 [Amol] DOC: Final touches
c30209d [Amol] DOC: Cleaning up
278e922 [Amol] DOC: Some cleaning up
e1877bf [Amol] DOC: Restructured the documentation
f394045 [Amol] DOC: Cleaned up the documentation
705e1a5 [Amol] DOC: Added an example of pitfalls when using astype

jreback committed:
TST: skip Fred / YahooOptions tests

jreback committed:
TST: split up test_merge

very small bug fix w.r.t. tz concatting

Author: Jeff Reback <jeff@reback.net>

Closes #13300 from jreback/merges and squashes the following commits:

727dfe9 [Jeff Reback] BUG: concat of same-tz needs string comparison for uniqueness
1560136 [Jeff Reback] TST: split up test_merge

jreback committed:
TST: reorg datetime with tz tests a bit

Author: Jeff Reback <jeff@reback.net>

Closes #13301 from jreback/ts2 and squashes the following commits:

2d8fbd4 [Jeff Reback] TST: reorg tests for datetime_with_tz construction

chris-b1 committed:
DOC: low_memory in read_csv

closes #5888,  xref #12686

Author: Chris <cbartak@gmail.com>

Closes #13293 from chris-b1/low-memory-doc and squashes the following commits:

daf9bca [Chris] DOC: low_memory in read_csv

ccronca committed:
ENH: support decimal argument in read_html #12907

closes #12907

Author: Camilo Cota <camilocot@gmail.com>
Author: Camilo Cota <ccota@riplife.es>

Closes #13272 from camilocot/issue-12907 and squashes the following commits:

0c15e37 [Camilo Cota] Remove bytes in decimal default value
111625f [Camilo Cota] ENH: support decimal argument in read_html #12907

jreback committed:
BUG: preserve join keys dtype

- closes #8596, preserve join keys dtype  - adds ``Index.where``
method for all Index types (like ``np.where/Series.where``), but
preserves dtypes

Author: Jeff Reback <jeff@reback.net>
Author: Mike Kelly <mtk@numeric.com>

Closes #13170 from jreback/merge2 and squashes the following commits:

0a267cf [Jeff Reback] BUG: preserve merge keys dtypes when possible
4173dbf [Mike Kelly] Preserve dtype in merge keys when possible

jreback committed:
COMPAT: windows test compat for merge, xref #13170

sinhrks committed:
TST: Make numpy_array test strict

assert_numpy_array_equal now checks input is np.ndarray
assert_almost_equal now checks inputs are the same class.

Author: sinhrks <sinhrks@gmail.com>

Closes #13311 from sinhrks/test_ndarray and squashes the following commits:

7ae4e93 [sinhrks] TST: Make numpy_array test strict

mortada committed:
DOC: remove references to deprecated numpy negation method

Author: Mortada Mehyar <mortada.mehyar@gmail.com>

Closes #13310 from mortada/deprecated_negation_in_docs and squashes the following commits:

3b46a1b [Mortada Mehyar] DOC: remove references to deprecated numpy negation method

sinhrks committed:
DOC: Fix read_stata docstring

Author: sinhrks <sinhrks@gmail.com>

Closes #13312 from sinhrks/stata_doc and squashes the following commits:

a9a7357 [sinhrks] DOC: Fix read_stata docstring

gfyoung committed:
BUG: Check for NaN after data conversion to numeric

Author: gfyoung <gfyoung17@gmail.com>

Closes #13314 from gfyoung/nan-check-post-numeric-conversion and squashes the following commits:

07f0538 [gfyoung] BUG: Check for NaN after data conversion to numeric

gfyoung committed:
TST: Parser tests refactoring

1) Moved no columns test from CParser-only to `common.py`
2) Moved erroneous placed skiprows tests into their proper place

Author: gfyoung <gfyoung17@gmail.com>

Closes #13319 from gfyoung/test-parsers-refactor and squashes the following commits:

bc1402e [gfyoung] TST: Parser tests refactoring

jennolsen84 committed:
COMPAT: do not upcast results to float64 when float32 scalar *+/- float64 array

closes #12388

Author: Jenn Olsen <jennolsen84@gmail.com>

Closes #12559 from jennolsen84/noevalupcast and squashes the following commits:

3f61252 [Jenn Olsen] do not upcast to float64 everytime

jreback committed:
TST: remove tests_tseries.py and distribute to other tests files

Author: Jeff Reback <jeff@reback.net>

Closes #13325 from jreback/reorg_tests and squashes the following commits:

c7e045e [Jeff Reback] TST: remove tests_tseries.py and distribute to other tests files

jreback committed:
BLD: increase clone depth

gfyoung committed:
ENH: add support for na_filter in Python engine

Title is self-explanatory.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13321 from gfyoung/python-engine-na-filter and squashes the following commits:

186fd34 [gfyoung] ENH: add support for na_filter in Python engine

jreback committed:
TST: more strict testing in lint.sh

Author: Jeff Reback <jeff@reback.net>

Closes #13334 from jreback/testing and squashes the following commits:

7653af9 [Jeff Reback] TST: more strict testing in lint.sh

pijucha committed:
BUG: Fix describe(): percentiles (#13104), col index (#13288)

closes #13104
closes #13288

Author: Piotr Jucha <pi.jucha@gmail.com>

Closes #13298 from pijucha/bug13104 and squashes the following commits:

9a6bd6e [Piotr Jucha] BUG: Fix describe(): percentiles (#13104), col index (#13288)

gfyoung committed:
ENH: Respect Key Ordering for OrderedDict List in DataFrame Init

Title is self-explanatory.  Closes #13304.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13309 from gfyoung/ordereddict-key-ordering-init and squashes the following commits:

4f311cc [gfyoung] ENH: Respect key ordering for OrderedDict list in DataFrame init

Commit:
BUG: Fix maybe_convert_numeric for unhashable objects

closes #13324

Author: Roger Thomas <roger.thomas@cremeglobal.com>

Closes #13326 from RogerThomas/fix_maybe_convert_numeric_for_unhashable_objects and squashes the following commits:

76a0738 [Roger Thomas] Fix maybe_convert_numeric for unhashable objects

jreback committed:
ENH: Series has gained the properties .is_monotonic*

Author: Jeff Reback <jeff@reback.net>

Closes #13336 from jreback/is_monotonic and squashes the following commits:

0a50ff9 [Jeff Reback] ENH: Series has gained the properties .is_monotonic, .is_monotonic_increasing, .is_monotonic_decreasing

jreback committed:
TST: computation/test_eval.py tests (slow)

closes #13338

Author: Jeff Reback <jeff@reback.net>

Closes #13339 from jreback/eval and squashes the following commits:

b2ee5e8 [Jeff Reback] TST: computation/test_eval.py tests (slow)

gfyoung committed:
BUG: Parse trailing NaN values for the Python parser

Fixes bug in which the Python parser failed to detect trailing `NaN`
values in rows

Author: gfyoung <gfyoung17@gmail.com>

Closes #13320 from gfyoung/trailing-nan-conversion and squashes the following commits:

590874d [gfyoung] BUG: Parse trailing NaN values for the Python parser

jorisvandenbossche committed:
DOC: fix comment on previous versions cythonmagic

Small thing I just noticed in the docs (the note on the other version
was not updated when the example was changed from cythonmagic ->
Cython)

Author: Joris Van den Bossche <jorisvandenbossche@gmail.com>

Closes #13343 from jorisvandenbossche/doc-cythonmagic and squashes the following commits:

902352c [Joris Van den Bossche] DOC: fix comment on previous versions cythonmagic

uwedeportivo committed:
Fix #13306: Hour overflow in tz-aware datetime conversions.

closes #13306

Author: Uwe Hoffmann <uwe@codemanic.com>

Closes #13313 from uwedeportivo/master and squashes the following commits:

be3ed90 [Uwe Hoffmann] whatsnew entry for issue #13306
1f5f7a5 [Uwe Hoffmann] Code Review jreback
82f263a [Uwe Hoffmann] Use vectorized searchsorted and tests.
a1ed5a5 [Uwe Hoffmann] Fix #13306: Hour overflow in tz-aware datetime conversions.

gfyoung committed:
ENH: Add support for compact_ints and use_unsigned in Python engine

Title is self-explanatory.    xref #12686 - I don't quite understand
why these are marked (if at all) as internal to the C engine only, as
the benefits for having these options accepted for the Python engine
is quite clear based on the documentation I added as well.
Implementation simply just calls the already-written function in
`pandas/parsers.pyx` - as it isn't specific to the `TextReader` class,
crossing over to grab this function from Cython (instead of
duplicating in pure Python) seems reasonable while maintaining that
separation between the C and Python engines.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13323 from gfyoung/python-engine-compact-ints and squashes the following commits:

95f7ba8 [gfyoung] ENH: Add support for compact_ints and use_unsigned in Python engine

gliptak committed:
BUG: Fix series comparison operators when dealing with zero rank numpy arrays

closes #13006

Author: Gábor Lipták <gliptak@gmail.com>

Closes #13307 from gliptak/seriescomp1 and squashes the following commits:

4967db4 [Gábor Lipták] Fix series comparison operators when dealing with zero rank numpy arrays

cswarth committed:
DOC: Add example usage to DataFrame.filter

Author: Chris Warth <cswarth@gmail.com>

Closes #12399 from cswarth/doc/df_filter and squashes the following commits:

f48e9ff [Chris Warth] DOC: Add example usage to DataFrame.filter

Commit:
DOC: Fixed a minor typo

Author: babakkeyvani <babak.keyvani@cpcc.edu>

Closes #13366 from bkeyvani/master and squashes the following commits:

029ade7 [babakkeyvani] DOC: Fixed a minor typo

gfyoung committed:
DOC: document doublequote in read_csv

Title is self-explanatory.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13368 from gfyoung/doublequote-doc and squashes the following commits:

f3e01fc [gfyoung] DOC: document doublequote in read_csv

gfyoung committed:
DEPR, DOC: Deprecate buffer_lines in read_csv

`buffer_lines` is not respected, as it is determined internally via a
heuristic involving `table_width` (see <a href="https://github.com/pyd
ata/pandas/blob/master/pandas/parser.pyx#L527">here</a> for how it is
computed).

Author: gfyoung <gfyoung17@gmail.com>

Closes #13360 from gfyoung/buffer-lines-depr-doc and squashes the following commits:

a72ecbe [gfyoung] DEPR, DOC: Deprecate buffer_lines in read_csv

chrish42 committed:
BUG: Make pd.read_hdf('data.h5') work when pandas object stored contained categorical columns

closes #13231

Author: Christian Hudon <chrish@pianocktail.org>

Closes #13359 from chrish42/gh13231 and squashes the following commits:

e839638 [Christian Hudon] Raise a better exception when the HDF file is empty and kwy=None.
611aa28 [Christian Hudon] Formatting fixes.
e7c8313 [Christian Hudon] Add changelog entry.
df10016 [Christian Hudon] Make logic that detects if there is only one dataset in a HDF5 file work when storing a dataframe that contains categorical data.
2f41aef [Christian Hudon] Tweak comment to be clearer.
b3a5773 [Christian Hudon] Add test that fails for GitHub bug #13231
02f90d5 [Christian Hudon] Use if-expression.

Commit:
DOC: remove obsolete cron job script (#13369)

* Typo correction

* removed deprecated script

jreback committed:
CLN: remove old skiplist code

Author: Jeff Reback <jeff@reback.net>

Closes #13372 from jreback/skiplist and squashes the following commits:

e05ea24 [Jeff Reback] CLN: remove old skiplist code

StephenKappel committed:
ENH: incorporate PR feedback; GH7271

StephenKappel committed:
ENH: incorporate PR feedback; GH7271

StephenKappel committed:
resolve merge conflict in rebasing of 7271-df-astype-dict

StephenKappel commented:
Ahh. Not sure how, but rebasing added everyone else's commits to this diff. Closing this PR and opening a clean one.

jreback commented:
you don't need to make a new PR just rebase and force push

StephenKappel commented:
I created a clean branch off of prod, cherry-picked my commits, and force pushed to this branch, but for some reason GitHub isn't recognizing it. Perhaps it's because I already closed the PR? But, I don't think I have access to reopen the PR...

StephenKappel commented:
Oh. It seems the PR can't be reopened because the branch has been subsequently force pushed :-(
